pandas/core/indexes/api.py

```diff
@@ -157,23 +158,42 @@ def _get_combined_index(
         index = ensure_index(index)
 
     if sort:
-        try:
-            array_sorted = safe_sort(index)
-            array_sorted = cast(np.ndarray, array_sorted)
-            if isinstance(index, MultiIndex):
-                index = MultiIndex.from_tuples(array_sorted, names=index.names)
-            else:
-                index = Index(array_sorted, name=index.name)
-        except TypeError:
-            pass
-
+        index = safe_sort_index(index)
     # GH 29879
     if copy:
         index = index.copy()
 
     return index
 
 
+def safe_sort_index(index: Index) -> Index:
+    """
+    Returns the sorted index
+
+    We keep the dtypes and the name attributes.
+
+    Parameters
+    ----------
+    index : an Index
+
+    Returns
+    -------
+    Index
+    """
+    try:
```

Comment:
following on from https://github.com/pandas-dev/pandas/issues/47329#issuecomment-1154087861

is our use of `safe_sort` here, actually safe? i.e. we don't reach any of the cases that raise in `safe_sort`) and the try is not needed?

Reply:
This gets adressed in #47331, we can probably remove the try there